burn-core/src/nn/loss/cross_entropy.rs

```diff
@@ -1,21 +1,107 @@
+#[derive(Config)]
+pub struct CrossEntropyLossConfig {
+    /// Create padded cross entropy.
+    ///
+    /// Prevents pad tokens from impacting loss calculation.
+    pad_tokens: Option<Vec<usize>>,
+
+    /// Create weighted cross-entropy.
+    ///
+    /// The loss of a specific sample will simply be given by: weight[y] * log(p(x)) * 1,
+    ///
+    /// # Pre-conditions
+    ///   - The order of the weight vector should correspond to the label integer assignment.
+    ///   - Targets assigned negative Int's will not be allowed.
+    weights: Option<Vec<f32>>,
+
+    /// Create cross-entropy with label smoothing.
+    ///
+    /// Hard labels {0, 1} will be changed to y_smoothed = y(1 - a) + a / nr_classes.
+    /// Alpha = 0 would be the same as default.
+    smoothing: Option<f32>,
+
+    /// Create binary cross-entrop.
+    ///
+    /// Will reduce computation by interpolating secondary label.  
+    /// Should only be given labels Tensor with {0, 1}.
+    /// If weights are also used, only two should be given.     
+    /// Cannot use pad_token at the same time.
+    #[config(default = false)]
+    binary: bool,
+}
+
+impl CrossEntropyLossConfig {
+    /// Initialize [Cross-entropy loss](CrossEntropyLoss).
+    pub fn init<B: Backend>(&self) -> CrossEntropyLoss<B> {
+        self.assertions();
+        CrossEntropyLoss {
+            pad_tokens: self.pad_tokens.clone(),
+            weights: self
+                .weights
+                .as_ref()
+                .map(|e| Tensor::<B, 1>::from_floats(e.as_slice())),
+            smoothing: self.smoothing,
+            binary: self.binary,
+        }
+    }
+
+    fn assertions(&self) {
+        if let Some(alpha) = self.smoothing {
+            assert!(
+                (0.0..=1.).contains(&alpha),
+                "Alpha of Cross-entropy loss with smoothed labels should be in interval [0, 1]. Got {}",
+                alpha
+            );
+        };
+        if let Some(weights) = self.weights.as_ref() {
+            assert!(
+                weights.iter().all(|e| e > &0.),
+                "Weights of cross-entropy have to be positive."
+            );
+        }
+        if let (Some(weights), true) = (self.weights.as_ref(), self.binary) {
+            assert!(
+                weights.len() == 2,
+                "Weights of binary cross-entropy should be of len 2. Got {}.",
+                weights.len()
+            );
+        }
+        if let (Some(_), true) = (&self.pad_tokens, self.binary) {
+            panic!(
+                "Pad tokens and binary option cannot be used at the same time in Cross Entropy."
+            );
+        }
+    }
+}
 
 /// Calculate the cross entropy loss from the input logits and the targets.
-#[derive(Clone, Debug)]
+#[derive(Module, Debug)]
 pub struct CrossEntropyLoss<B: Backend> {
-    pad_index: Option<usize>,
-    backend: PhantomData<B>,
+    pad_tokens: Option<Vec<usize>>,
+    weights: Option<Tensor<B, 1>>,
```

Comment:
I would put the weights public, so we could print them, or use a tensor from someplace else.